File: design-patterns/ts/factory/transport-factory-method/road-transport-factory.ts
// road-transport-factory.ts

import Car from "../transport/car";
import TransportFactory from "./transport-factory";
import Bus from "./bus";
import Bike from "./bike";
import Transport from "./transport";

class RoadTransportFactory extends TransportFactory {
    getTransport(name: string): (Transport | null) {
        if (name.toLowerCase() === "car") {
            return new Car();
        }

        if (name.toLowerCase() === "bike") {
            return new Bike();
        }

        if (name.toLowerCase() === "bus") {
            return new Bus();
        }

        return null;
    }
}

export default RoadTransportFactory;

File: design-patterns/go/factory/transport-method/road_transport_factory.go
// road_transport_factory.go

package main

import (
	"errors"
	"strings"
)

type RoadTransportFactory struct {
	TransportFactory
}

func NewRoadTransportFactory() (roadTransportFactory *RoadTransportFactory) {
	roadTransportFactory = &RoadTransportFactory{TransportFactory{}}
	roadTransportFactory.TransportFactory.TransportFactoryInterface = roadTransportFactory
	return
}

func (roadTransportFactory *RoadTransportFactory) GetTransport(name string) (Transport, error) {
	if strings.ToLower(name) == "car" {
		return NewCar(), nil
	}

	if strings.ToLower(name) == "bike" {
		return NewBike(), nil
	}

	if strings.ToLower(name) == "bus" {
		return NewBus(), nil
	}

	return nil, errors.New("unknown identifier for RoadTransportFactory")
}
